include/grpc++/channel_filter.h

```diff
@@ -0,0 +1,211 @@
+
+ protected:
+  ChannelData() {}
+};
+
+namespace internal {
+
+// Defines static members for passing to C core.
+template<typename ChannelDataType, typename CallDataType>
+class ChannelFilter {
+  static const size_t call_data_size = sizeof(CallDataType);
+
+  static void InitCallElement(
+      grpc_exec_ctx *exec_ctx, grpc_call_element *elem,
+      grpc_call_element_args *args) {
+    CallDataType* call_data = elem->call_data;
+    *call_data = CallDataType();
+  }
+
+  static void DestroyCallElement(
+      grpc_exec_ctx *exec_ctx, grpc_call_element *elem,
+      const grpc_call_stats *stats, void *and_free_memory) {
+    CallDataType* call_data = elem->call_data;
+    // Can't destroy the object here, since it's not allocated by
+    // itself; instead, it's part of a larger allocation managed by
+    // C-core.  So instead, we call the Destroy() method.
+    call_data->Destroy();
+  }
+
+  static void StartTransportStreamOp(
+      grpc_exec_ctx *exec_ctx, grpc_call_element *elem,
+      grpc_transport_stream_op *op) {
+    CallDataType* call_data = elem->call_data;
+    call_data->StartTransportStreamOp(exec_ctx, op);
+  }
+
+  static void SetPollsetOrPollsetSet(
+      grpc_exec_ctx *exec_ctx, grpc_call_element *elem,
+      grpc_polling_entity *pollent) {
+    CallDataType* call_data = elem->call_data;
+    call_data->SetPollsetOrPollsetSet(exec_ctx, pollent);
+  }
+
+  static char* GetPeer(
+      grpc_exec_ctx *exec_ctx, grpc_call_element *elem) {
+    CallDataType* call_data = elem->call_data;
+    return call_data->GetPeer(exec_ctx);
+  }
+
+  static const size_t channel_data_size = sizeof(ChannelDataType);
+
+  static void InitChannelElement(
+      grpc_exec_ctx *exec_ctx, grpc_channel_element *elem,
+      grpc_channel_element_args *args) {
+    ChannelDataType* channel_data = elem->channel_data;
+    *channel_data = ChannelDataType();
+  }
+
+  static void DestroyChannelElement(
+      grpc_exec_ctx *exec_ctx, grpc_channel_element *elem) {
+    ChannelDataType* channel_data = elem->channel_data;
+    // Can't destroy the object here, since it's not allocated by
+    // itself; instead, it's part of a larger allocation managed by
+    // C-core.  So instead, we call the Destroy() method.
+    channel_data->Destroy();
+  }
+
+  static void StartTransportOp(
+      grpc_exec_ctx *exec_ctx, grpc_channel_element *elem,
+      grpc_transport_op *op) {
+    ChannelDataType* channel_data = elem->channel_data;
+    channel_data->StartTransportOp(exec_ctx, op);
+  }
+};
+
+struct FilterRecord {
+  grpc_channel_stack_type stack_type;
+  int priority;
+  grpc_channel_filter filter;
+};
+extern std::vector<FilterRecord>* channel_filters;
+
+void ChannelFilterPluginInit();
+void ChannelFilterPluginShutdown() {}
+
+}  // namespace internal
+
+// Registers a new filter.
+// Must be called by only one thread at a time.
+template<typename ChannelDataType, typename CallDataType>
+void RegisterChannelFilter(const char* name,
```

Comment:
We might want to add the argument:
```
std::function<bool(grpc_channel_args*)> should_this_filter_be_included
```
to allow runtime decisions to be made about this filter